Fix: Loading packages from custom URLs (part 1)

rth opened the pull request:
Fixes the first part of https://github.com/iodide-project/pyodide/issues/121

This PR allows to load packages from user defined URLs (without handling dependencies for now).  A few details on the current implementation,
 - the URL is assumed to start with http or https, the package name is inferred from the URL using the template (`https://some_path/{package_name}.js`)
 - packages names are assumed to composed of letters, numbers and/or `-`, `_`
 - if the provided URI doesn't look like a URL or a package name, an error is sent to the `console.log` and no loading takes place.
 - if one attempts to load the same package from two separate sources (e.g. a URL and `package.json`) an error will be raised. Loading the same package twice from the same URI does nothing.

rth commented on src/pyodide.js:
```diff
@@ -11,20 +18,53 @@ var languagePluginLoader = new Promise((resolve, reject) => {
-      const package = queue.pop();
-      if (!loadedPackages.has(package)) {
-        toLoad.add(package);
+      let package_uri = queue.pop();
```
Using the term `URI` here and elsewhere to indicate that it could be a URL or a package name directly.

rth commented on src/pyodide.js:
```diff
@@ -11,20 +18,53 @@ var languagePluginLoader = new Promise((resolve, reject) => {
+      const package = _uri_to_package_name(package_uri);
+
+      if (package == null) {
+        console.log(`Invalid package name or URI '${package_uri}'`);
```
So currently there is no mechanism to raise an error in iodide short of printing something to `console.log`?

mdboom replied:
You can also `throw new Error(...)` if it really is fatal.  That will surface up to the Iodide interface, or outside of Iodide will end up in the console.

rth commented on src/pyodide.js:
```diff
@@ -11,20 +18,53 @@ var languagePluginLoader = new Promise((resolve, reject) => {
+        package_uri = 'packages.json';
+      }
+
+      console.log(`Loading ${package} from ${package_uri}`);
```
Added this for debugging, but it could be removed.

rth commented on test/test_package_loading.py:
```diff
@@ -0,0 +1,50 @@
+
+
+@pytest.fixture
+def static_web_server():
```
We may want to standarize this with the webserver started in `tests/conftest.py`. In particular, if I understand correctly that one uses multiprocessing on some free port. Maybe moving both under some pytest fixture in `conftest.py` might make sense.

mdboom replied:
I thought about using a fixture, but in this case, I truly wanted it to be a single webserver for *all* tests, so I didn't see much value.  

I've always preferred using multiprocessing for this because then the GIL isn't in contention when communicating between the webserver and the test process -- though in this case all the tests are in selenium (and thus a separate process anyway), so maybe it doesn't matter.

Maybe the thing to do is add a fixture to `conftest.py` that just returns the url and port to the singleton webserver:

```
@pytest.fixture
def web_server():
    yield '127.0.0.1', PORT
```

Then we wouldn't need this one.

rth replied:
>  in this case, I truly wanted it to be a single webserver for all tests, so I didn't see much value.

I think a `pytest.fixture(scope='session')` would do just that?

Agreed that since we have free CPU cores, we might make use of them without worrying about the GIL.

> Maybe the thing to do is add a fixture to conftest.py that just returns the url and port to the singleton webserver:

Here, initially wanted to create a separate server with a different base URL / PORT to be sure that this is working as expected. You are right that re-using a single server might be sufficient though.

rth commented on test/test_package_loading.py:
```diff
@@ -0,0 +1,50 @@
+    build_dir = Path(__file__).parent.parent / 'build'
+    os.chdir(build_dir)
+    try:
+        url, port = ('127.0.0.1', 8888)
```
This assumes that port `8888` is unused, not sure to what extent that may be an issue.

mdboom replied:
It becomes an issue when the test process doesn't shut down cleanly -- then the port remains reserved by the system for a few seconds and can't be opened again.

rth commented on src/pyodide.js:
```diff
@@ -2,6 +2,13 @@
+// Regexp for validating package name and URI
+var package_name_regexp = '[a-z0-9_\-]+'
+var package_uri_regexp = new RegExp(
+     '^https?://.*?(' + package_name_regexp + ').js$', 'i');
```
Initially I though that e.g. `file://` would also work here, but it doesn't seem to be valid. Not sure what should be the approach to load `.js` files from local system.

mdboom replied:
Let's add this to the docs.

mdboom commented on .circleci/config.yml:
```diff
@@ -69,6 +69,9 @@ jobs:
             make test
 
+      - store_artifacts:
+          path: /home/circleci/repo/build/
```
Cool.  I didn't realize this would be so easy.  This reminds me I need to get auto-deployment working again.

mdboom commented on src/pyodide.js:
```diff
@@ -11,20 +18,53 @@ var languagePluginLoader = new Promise((resolve, reject) => {
+        console.log(`Invalid package name or URI '${package_uri}'`);
+        break;
+      } else if (package == package_uri) {
+        package_uri = 'packages.json';
```
Maybe call this `__builtin__` or something?  It wasn't clear to me that this is just a special token for builtin packages, and that `packages.json` is loaded elsewhere (never as the result of this value).

rth replied:
Called this "default channel", hope it's OK, as I think it's a bit confusing say e.g.
```
Installing numpy from __builtin__
```

mdboom replied:
Ah.  I missed that it was displayed to the user.  That makes sense.

rth commented:
BTW, here we are exposing build artifacts in Circle CI (they can be seen under the "Artifacts" tab), so if we open, for instance, `python.html` from the latest build,

https://255-122663163-gh.circle-artifacts.com/0/home/circleci/repo/build/python.html

and load `pyiodide.js` from the same base URL, this feature can be tested "live" in that notebook :)

rth commented:
> Finally, what are the unique keys to determine if a package is already loaded? It should probably be the resulting python package name (i.e. matplotlib), not the full URL to where the package came from, since it's impossible to load two python packages called matplotlib at the same time, anyway. I don't think the .js files contain that information already, so we'll need to add that as well.

For now, I am just inferring the unique key with from the URL as that allows to do all the logic without loading the file, but it's true that including that in .js would be more robust. Maybe this could be done in a follow up PR though? Here I tried to get the functionality with minimal changes..

mdboom commented on test/conftest.py:
```diff
@@ -46,7 +46,11 @@ def __init__(self):
     def logs(self):
-        return self.driver.execute_script("return window.logs")
+        logs = self.driver.execute_script("return window.logs")
+        return '\n'.join(str(x) for x in logs)
```
This makes it harder to assert that an entire line appears exactly in the log.  See `test_python.py::test_init`, for example.

I think this should remain as returning a list, but you could add a `logs_as_string` helper function.

rth replied:
I was wondering about that. Will create a separate helper function.

The question is though how often we will match a substring as opposed to a full line. If more often, `some_line in selenium.logs.splitlines()` might be not so bad, compared to,
```py
some_substr in selenium.logs_as_string
# together with
some_line in selenium.logs
```
I am more used for logs to be a string  rather than a list (even if it means doing back and forth conversions internally) but it is indeed subjective..

mdboom commented on src/pyodide.js:
```diff
@@ -2,6 +2,13 @@
  */
 
+// Regexp for validating package name and URI
+var package_name_regexp = '[a-z0-9_\-]+'
```
Also, they *can* include capital letters (but rarely do due to PEP8).

This [SO comment](https://stackoverflow.com/questions/5474008/regular-expression-to-confirm-whether-a-string-is-a-valid-identifier-in-python) may be helpful.  I have no idea if the proposed regex there translates easily into a Javascript regex.

rth replied:
Well `python-dateutil` is a package name that is currently used... As far as I understand here package name would differ from the import name, and I think this situation makes sense since multiple packages can have the same import name (e.g. https://github.com/seomoz/simhash-py and https://github.com/leonsim/simhash which both import as `simhash`).

Here the regexp doesn't include capitals, but it is initialized with the [`i` flag](https://github.com/leonsim/simhash) to ignore case.

mdboom replied:
You are right.  Thanks for clearing my confusion.  In this case we need to use the import name, since you can't have two packages with the same import name installed at the same time.  That makes using the filename as a proxy for import name a bit difficult, of course.

I missed the ignore case flag.  That seems fine, then.

rth commented:
Thanks for the review @mdboom!  Will address your comments tomorrow..

rth commented:
I think I have addressed all review comments except for,

> In this case we need to use the import name, since you can't have two packages with the same import name installed at the same time. That makes using the filename as a proxy for import name a bit difficult, of course.

I agree, that what happens when two packages with the same import name are installed is an important point. I am a bit concerned about using import names here as unique identifier, as,
 - a package can install multiple modules and thus have multiple import names (e.g. [setuptools](https://github.com/conda-forge/setuptools-feedstock/blob/f8ce0a13b947e62ccd0fe974c42a9ccf86b44720/recipe/meta.yaml#L35-L37))
 - if we see `pyodide.loadPackage` as analogy for pip/conda install, I think it would make sense to use package names (instead of import names): i.e. `pyodode.loadPackage('scikit-learn')` instead of `pyodide.loadPackage('sklearn')` and same would apply to URLs (to be consistent with `.tar.gz` URLs etc)
 - having to worry (and track it) import names  e.g. in `meta.yaml` would make some things harder such as future hacks to install from PyPi (cf [wiki](https://github.com/iodide-project/pyodide/wiki/Packaging)).
 - in this part of the code, I don't really have access to import names (particularly for URLs): all I can do is to go to the URL, infer some information from it's content, and load the JS.

So my questions is can we keep this in it's current form (and use package name as identifier), then in later PR add some validation for this case, when we create the filesystem overlay and have access to all the data? Can we not just error at that point with the message that import path already exists? I'm not really sure what happens when one pip installs two packages with the same import name, does it silently overwrite it, or prints some warnings? Could we not make something similar happen here which might be simpler than tracking import names?
